### DISCUSSION TITLE
Run neurodesktop docker container from the littlest jupyterhub?

### DISCUSSION THREAD

**Discussion by UppsalaBrainBehaviorLab** [neurodesk/neurodesk.github.io-hugo#395] [Q&A] [answered]
Hello,
I am working on setting up a neuroimaging analysis server on a linux server (ubuntu 22:04) with multiple users (10-15). I was thinking that running neurodesktop as a docker container from jupyterhub might be a good option, as it would allow users to access the desktop through their web browser and each user would get their own instance of neurodesktop. I am using dockerspawner to try to start the image vnmd/neurodesktop:2023-11-28. But I am running into problems with the setup.

I can run neurodesktop as a docker container on the linux server and connect using the web browser and the neurodeskapp, but it's the part of starting it from jupyterhub that is failing. I get no error, it just seems to start some basic notebook.

This is my config.py located in /opt/tljh/config/jupyterhub_config.d/. The other notebooks launch OK.
```
c.JupyterHub.spawner_class = 'dockerspawner.DockerSpawner'
c.DockerSpawner.allowed_images = ['jupyter/datascience-notebook:r-4.0.3', 'jupyter/datascience-notebook:r-3.6.3', 'vnmd/neurodesktop:2023-11-28']

from jupyter_client.localinterfaces import public_ips
c.JupyterHub.hub_ip = public_ips()[0]

c.DockerSpawner.name_template = '{prefix}-{username}-{servername}'
```

First of all, I don't know if this is even possible. If it is, can you point me in the right direction? Or do you know of any tutorial or guide on how to accomplish this? 

Thanks!
Andreas

**Comment by stebo85** [comment]
Dear @UppsalaBrainBehaviorLab,

that could work but we haven't tried this particular setup yet. What we have tried and many people got to work well is to use Kubernetes to run Jupyterlab (https://z2jh.jupyter.org/en/stable/) and then run the Neurodesk image in there. Would that be an option for you? It seems that tljh seems like a great approach for a single node setup (https://tljh.jupyter.org/en/latest/topic/whentouse.html) and I am not sure why what you tried doesn't work yet. 

Can you share more details what you exactly tried and what happens? Is the change in the jupyterhub_config.d/config.py the only one you made? What you do you mean with "it just seems to start some basic notebook" - how does this notebook look like? Is this inside the neurodesk docker container?

One other problem I anticipate is that you might not be able to run cvmfs inside the container (except when running as privileged). So I would recommend to install CVMFS on the host: https://www.neurodesk.org/docs/getting-started/neurocontainers/cvmfs/ - our image recognizes on startup if this directory is there, but you will need to provide a mount command to the docker spawner. 

Kind regards
Steffen

**Comment by UppsalaBrainBehaviorLab** [comment]
Hello @stebo85,

Thanks for your reply. I first tried tljh since we are only using one node and I thought Kubernetes would be too much hassle without any gains.

I installed tljh as according to their instructions at https://tljh.jupyter.org/en/latest/install/custom-server.html
`curl -L https://tljh.jupyter.org/bootstrap.py | sudo -E python3 - --admin <adminusername>
`

I then added a config.py to the jupyterhub_config.d directory. I tried some different versions without success, this is the simplest:
```
c.JupyterHub.spawner_class = 'dockerspawner.DockerSpawner'
c.DockerSpawner.image = 'vnmd/neurodesktop:2023-11-28'

from jupyter_client.localinterfaces import public_ips
c.JupyterHub.hub_ip = public_ips()[0]

```

I don't have much experience with jupyterhub or lab, but to me it looks like it starts some sort of default notebook. There is no mentionof neurodesktop in the logfiles from journalctl -u jupyterhub or journalctl -u jupyter-USER. I see this:
![image](https://github.com/NeuroDesk/neurodesk.github.io/assets/72663784/89c35b35-831f-4502-bb1f-e4fbcface9f8)

Any thoughts appreciated.

Cheers,
Andreas

**Reply by stebo85** [reply]
Dear @UppsalaBrainBehaviorLab - Ok, I will try to reproduce your setup and see what I can find. Where did you find that DockerSpawner is actually compatible with TLJH? I couldn't find anything about DockerSpawner in their documentation.

Thank you
Steffen

**Reply by UppsalaBrainBehaviorLab** [reply]
Dear @stebo85,

Yes, the dockerspawner works with TLJH. If I add this to my config file (inspired by https://github.com/pennchildlanglab/tljh-datascience/blob/main/tljh_datascience.py) 
```
c.JupyterHub.spawner_class = 'dockerspawner.DockerSpawner'
c.DockerSpawner.image_whitelist = ['jupyter/datascience-notebook:r-4.0.3', 'vnmd/neurodesktop:2023-11-28']

from jupyter_client.localinterfaces import public_ips
c.JupyterHub.hub_ip = public_ips()[0]

c.DockerSpawner.name_template = '{prefix}-{username}-{servername}'
```
I can pull and start the datascience-notebook image. But the neurodesktop one does not work. I am thinking it might have something to do with IPs or ports, but am not seeing exactly how to get to this to work.

Thanks!
Andreas

**Reply by stebo85** [reply]
Dear @UppsalaBrainBehaviorLab 

I was able to reproduce your setup now and get the same problem. I currently think the problem is that tljh-datascience doesn't actually load the containers from the list specified. It's always loading up their default jupyter/datascience-notebook:r-4.0.3 no matter what's specified.

I am hoping I can find out why that is the case, but it doesn't seem to be a problem with our Neurodesk image.

**Reply by stebo85** [reply]
Dear @UppsalaBrainBehaviorLab - a quick update on this: The tlhj_datascience plugin had a couple of problems and bugs. We started from this and fixed these issues, so Neurodesk is now running. We are doing a few more final tests and then you can try it out :)

**Reply by UppsalaBrainBehaviorLab** [reply]
Dear @stebo85 , That is super! Thanks a lot for pursuing this.

**Reply by stebo85** [reply, marked as answer]
Dear @UppsalaBrainBehaviorLab - it took a while, but we finally got neurodesk to work with a TLJH plugin. Can you test this out here https://github.com/NeuroDesk/tljh-neurodesk and let us know how far you get?

It's not as nice of a setup as using Kubernetes, but it might work well enough for your usecase :)

**Comment by UppsalaBrainBehaviorLab** [comment]
Thanks so much for the work @stebo85! I will test it and report back.

**Comment by UppsalaBrainBehaviorLab** [comment]
Dear @stebo85,

I tried the install script and got it to work with a minor edit. Thanks a lot!

First, I received this error when running the install command:
```
<---lines removed--->
Status: Downloaded newer image for vnmd/neurodesktop:latest
docker.io/vnmd/neurodesktop:latest
Traceback (most recent call last):
  File "/usr/bin/docker-compose", line 33, in <module>
    sys.exit(load_entry_point('docker-compose==1.29.2', 'console_scripts', 'docker-compose')())
  File "/usr/lib/python3/dist-packages/compose/cli/main.py", line 81, in main
    command_func()
  File "/usr/lib/python3/dist-packages/compose/cli/main.py", line 200, in perform_command
    project = project_from_options('.', options)
  File "/usr/lib/python3/dist-packages/compose/cli/command.py", line 60, in project_from_options
    return get_project(
  File "/usr/lib/python3/dist-packages/compose/cli/command.py", line 152, in get_project
    client = get_client(
  File "/usr/lib/python3/dist-packages/compose/cli/docker_client.py", line 41, in get_client
    client = docker_client(
  File "/usr/lib/python3/dist-packages/compose/cli/docker_client.py", line 124, in docker_client
    kwargs = kwargs_from_env(environment=environment, ssl_version=tls_version)
TypeError: kwargs_from_env() got an unexpected keyword argument 'ssl_version'
```

I identified this as originating from the docker-compose command in the tljh_neurodesk.py, line 15. I am not sure this is the best way, but it was solved by changing docker-compose to docker compose:
`docker compose up -d""")`

**Reply by stebo85** [reply]
Dear @UppsalaBrainBehaviorLab - thank you for reporting this :)

We didn't see this error in our testing, but we think we have an idea where it comes from.

Could it be that you had docker-ce installed already before you ran the script? We assume the debian docker package and it looks like these different docker version have a slightly different syntax for the compose command?

**Reply by UppsalaBrainBehaviorLab** [reply]
@stebo85 Yes, this would explain the issue. We had docker-ce installed on the system. Thanks for the explanation.

**Reply by stebo85** [reply]
great, we added a workaround for this!
